### PR TITLE
W-10999318: Service Schedule "Add More Sessions" Wizard

### PR DESCRIPTION
**Pull request by ptewson-sfdo**
# Critical Changes

# Changes

# Issues Closed
- [W-10999318](https://gus.lightning.force.com/a07EE00000uxxW1YAI)

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [ ] Any net new LWC work has Sa11y tests & 50% or above lines JEST test coverage  
- [x] CRUD/FLS is enforced in Apex
- [x] Permission sets are updated to account for CRUD|FLS|Tab|Classes
- [x] Field sets are updated to account for new fields
- [ ] UX approval or UX not necessary
- [ ] Link the pull request and work item by PR comment and Chatter post respectively, e.g. GUS: W-0000000: Work Name (https://github.com/SalesforceFoundation/PMM/pull/303)
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] PR contains draft release notes
- [ ] QE story level testing completed

### CODE REVIEW

File: force-app/main/default/classes/ServiceSessionSelector.cls [hunk 1]
```diff
@@ -15,6 +15,42 @@ public with sharing class ServiceSessionSelector {
...
+
+        Set<String> fields = new Set<String>();
+        fields.add(String.valueOf(ServiceSession__c.Id)); // Name is required
+        fields.add(String.valueOf(ServiceSession__c.Name)); // Name is required
+        fields.add(String.valueOf(ServiceSession__c.Status__c)); // Status is required
+        fields.add(String.valueOf(ServiceSession__c.SessionStart__c)); // Start date is used to group the records
+        fields.add(String.valueOf(ServiceSession__c.SessionEnd__c)); // Start date is used to group the records
+        fields.add(String.valueOf(ServiceSession__c.ServiceSchedule__c)); // Start date is used to group the records
+
+        QueryBuilder queryBuilder = new QueryBuilder()
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceSessionSelector.cls:35]: If we know all of the fields ahead of time that we are querying for, do we need to use the query builder or can we just use SELECT...?

File: force-app/main/default/lwc/participantSelector/participantSelector.js [hunk 2]
```diff
@@ -270,13 +270,40 @@ export default class ParticipantSelector extends LightningElement {
 
         this.previouslySelectedEngagements = [...this.previouslySelectedEngagements];
         this.availableEngagementRows.forEach(row => {
-            if (
-                this.existingContactIds.includes(
+            let previouslySelectedEngagementIndex = this.previouslySelectedEngagements.findIndex(
```
Comment (jjbennett) [inline force-app/main/default/lwc/participantSelector/participantSelector.js:273]: Are these changes still needed?

File: force-app/main/default/lwc/reviewSessions/reviewSessions.js [hunk 3]
```diff
@@ -69,11 +70,11 @@ export default class ReviewSessions extends LightningElement {
 
         this._serviceSessions = [...this._serviceScheduleModel.serviceSessions];
 
-        if (!this._serviceSessions.length) {
-            this.getSessions();
-        } else {
-            this.handleDispatchLoadedEvent();
-        }
+        //if (!this._serviceSessions.length) {
```
Comment (jjbennett) [inline force-app/main/default/lwc/reviewSessions/reviewSessions.js:73]: Remove comments?

File: force-app/main/default/lwc/reviewSessions/reviewSessions.js [hunk 4]
```diff
@@ -201,14 +212,15 @@ export default class ReviewSessions extends LightningElement {
             },
             {
                 label: "",
+                cellAttributes: { class: { fieldName: "alreadyCreatedCSSClass" } },
```
Comment (jjbennett) [inline force-app/main/default/lwc/reviewSessions/reviewSessions.js:215]: My only hesitation on using the background color is that we do not do the same thing with the existing program engagements.

File: force-app/main/default/lwc/serviceScheduleCreator/serviceScheduleCreator.js [hunk 5]
```diff
@@ -308,16 +355,45 @@ export default class ServiceScheduleCreator extends NavigationMixin(LightningEle
...
-        this.serviceScheduleModel.serviceSchedule[
-            SERVICE_FIELD.fieldApiName
-        ] = this.serviceId;
+        if (this.serviceId) {
+            this.serviceScheduleModel.serviceSchedule[
+                SERVICE_FIELD.fieldApiName
+            ] = this.serviceId;
+        }
+
+        let flattenedSelectedEngagments = this.serviceScheduleModel.selectedEngagements.map(
```
Comment (jjbennett) [inline force-app/main/default/lwc/serviceScheduleCreator/serviceScheduleCreator.js:380]: Is this change still needed?

**Review by jjbennett — COMMENTED** [review]
I just did a quick run through of the code and have some comments and questions. The only other thought I had in the chatter for the work item is around the modal header. It still says New Service Schedule. I'm guessing we want to make that match the button label. Let me know if you have any questions. Thanks!

File: force-app/main/default/classes/ServiceScheduleModel.cls [hunk 6]
```diff
@@ -133,6 +202,53 @@ public with sharing class ServiceScheduleModel {
         );
     }
 
+    public void addServiceSessions(List<ServiceSession__c> newServiceSessions) {
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleModel.cls:205]: Probably not! I can clear this out. RIP service session key!

File: force-app/main/default/classes/ServiceScheduleModel.cls [hunk 7]
```diff
@@ -103,10 +107,75 @@ public with sharing class ServiceScheduleModel {
...
+                PrimaryServiceProvider__c,
+                ServiceScheduleEndDate__c,
+                ServiceScheduleEnds__c,
+                Service__c
+            FROM ServiceSchedule__c
+            WHERE Id = :serviceScheduleId
+            LIMIT 1
+        ];
+
+        if (serviceSessions == null) {
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleModel.cls:127]: Take a look at it now - I think this is in line with your suggestion but let me know if you want me to separate it further.

File: force-app/main/default/classes/ServiceScheduleDomain.cls [hunk 8]
```diff
@@ -36,8 +41,16 @@ public with sharing class ServiceScheduleDomain {
     }
 
     private void insertRelatedRecords(List<SObject> sObjects) {
-        validateCreateAccess(sObjects);
-        insert sObjects;
+        List<SObject> sObjectsNotYetInserted = new List<SObject>();
+        for (SObject sObj : sObjects) {
+            if (sObj.Id == null) {
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleDomain.cls:46]: This feels a little off to me, we shouldn't be sending records with Ids to an insert method.

File: force-app/main/default/classes/ServiceScheduleDomain_TEST.cls [hunk 9]
```diff
@@ -43,6 +43,37 @@ private with sharing class ServiceScheduleDomain_TEST {
...
+            LIMIT 1
+        ];
+
+        String originalName = originalSchedule.Name;
+
+        ServiceScheduleModel model = new ServiceScheduleModel(originalSchedule.Id, null);
+        model.serviceSchedule.Name = originalName + ' (Updated)';
+
+        Test.startTest();
+        domain.insertModel(model);
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleDomain_TEST.cls:60]: Should this be changed to upsertModel?

File: force-app/main/default/classes/ServiceScheduleModel.cls [hunk 10]
```diff
@@ -81,6 +85,10 @@ public with sharing class ServiceScheduleModel {
         'name' => fieldSetService.getFieldForLWC(
             Schema.ServiceSession__c.Name.getDescribe()
         ),
+        'id' => fieldSetService.getFieldForLWC(Schema.ServiceSession__c.Id.getDescribe()),
+        'createdDate' => fieldSetService.getFieldForLWC(
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleModel.cls:89]: I don't think you need this anymore?

File: force-app/main/default/classes/ServiceScheduleSelector_TEST.cls [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    private static void createScheduleData() {
+        TestDataFactory.generateServiceData();
+        TestDataFactory.generateServiceScheduleData('Daily', System.today());
+    }
+
+    @IsTest
+    private static void testGetServiceScheduleById() {
+        ServiceSchedule__c expectedSchedule = [
+            SELECT
+                Id,
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleSelector_TEST.cls:22]: If we are only referencing the Id and the Name in the test, do we need to query the rest of the fields?

File: force-app/main/default/lwc/reviewSessions/reviewSessions.js [hunk 12]
```diff
@@ -33,6 +33,7 @@ export default class ReviewSessions extends LightningElement {
     sessionNameLabel;
     sessionStartLabel;
     sessionEndLabel;
+    sessionIdLabel;
```
Comment (jjbennett) [inline force-app/main/default/lwc/reviewSessions/reviewSessions.js:36]: Is this needed?

File: force-app/main/default/lwc/serviceScheduleCreator/serviceScheduleCreator.js-meta.xml [hunk 13]
```diff
@@ -1,5 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
-    <apiVersion>49.0</apiVersion>
-    <isExposed>false</isExposed>
+    <apiVersion>54.0</apiVersion>
+    <isExposed>true</isExposed>
+    <targets>
+        <target>lightning__RecordAction</target>
```
Comment (jjbennett) [inline force-app/main/default/lwc/serviceScheduleCreator/serviceScheduleCreator.js-meta.xml:6]: Do we still need this if you added the wrapper?

File: force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js-meta.xml [hunk 14]
```diff
@@ -1,5 +1,13 @@
...
-    <apiVersion>51.0</apiVersion>
-    <isExposed>false</isExposed>
+    <apiVersion>54.0</apiVersion>
+    <isExposed>true</isExposed>
+    <targets>
+        <target>lightning__RecordAction</target>
+    </targets>
+    <targetConfigs>
+        <targetConfig targets="lightning__RecordAction">
+            <actionType>ScreenAction</actionType>
```
Comment (jjbennett) [inline force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js-meta.xml:10]: Or this?

File: force-app/main/default/quickActions/ServiceSchedule__c.Add_More_Sessions.quickAction-meta.xml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<QuickAction xmlns="http://soap.sforce.com/2006/04/metadata">
+    <actionSubtype>ScreenAction</actionSubtype>
+    <label>Add More Sessions</label>
+    <lightningWebComponent>serviceScheduleAddSessionsWrapper</lightningWebComponent>
```
Comment (jjbennett) [inline force-app/main/default/quickActions/ServiceSchedule__c.Add_More_Sessions.quickAction-meta.xml:5]: Instead of Wrapper what about being specific with Quick Action?

File: force-app/main/default/classes/ServiceScheduleDomain.cls (hunk 8 above)
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleDomain.cls:46]: Do you have a suggestion? We could filter out pre-existing records that have Ids outside of this method, or possibly re-work this as an upsert?

File: force-app/main/default/classes/ServiceScheduleDomain_TEST.cls (hunk 9 above)
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleDomain_TEST.cls:60]: I believe so (although, with these updates mis-naming is going to run rampant throughout the entire "Service Schedule Creator" framework)

**Review by jjbennett — COMMENTED** [review]
I did a quick read over of the code, some naming suggestions but cool if you don't agree with them. I'll do a more thorough review tomorrow. Thanks!

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 16]
```diff
@@ -365,8 +454,7 @@ export default class NewServiceSchedule extends LightningElement {
         if (!event.detail.value) {
             this.dateFields.end.value = null;
         } else {
-            this.dateFields.start.value = event.detail.value;
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:368]: this.dateFields.start.value gets the value of the passed-in argument within setFirstSessionStartTimeAndEndTime(), so it's redundant as far as I can tell.

File: force-app/main/default/classes/ServiceScheduleService.cls [hunk 17]
```diff
@@ -17,19 +17,37 @@ public with sharing class ServiceScheduleService {
...
+        }
+        return ssModel;
     }
 
     public ServiceScheduleModel persist(ServiceScheduleModel model) {
         Savepoint savepoint = Database.setSavepoint();
 
         try {
-            domain.insertModel(model);
+            domain.commitChangesToModel(model);
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleService.cls:50]: What do you think would be the ideal arrangement ... this method goes into ServiceScheduleService?

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 18]
```diff
@@ -192,6 +199,13 @@
         <shortDescription>Supporting message for Attendance empty state</shortDescription>
         <value>After participants are added to this Service Schedule, track their attendance here.</value>
     </labels>
+    <labels>
+        <fullName>Add_Sessions_Message</fullName>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Supporting message for Adding Sessions to an existing schedule.</shortDescription>
+        <value>You are updating an existing schedule any changes you make here will be saved to the schedule record. Please note that you will only be able to add sessions after the most current existing sessions.</value>
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/labels/CustomLabels.labels-meta.xml:207]: I actually missed it previously, but saw it earlier today. I will update these labels, and thanks for sussing that out!

File: force-app/main/default/classes/ServiceScheduleService.cls (hunk 17 above)
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/ServiceScheduleService.cls:45]: Yea I think that would have been a better spot.

File: force-app/main/default/classes/ServiceScheduleSelector_TEST.cls [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        Test.startTest();
+        System.runAs(TestUtil.getTestUser()) {
+            ServiceScheduleSelector scheduleSelector = new ServiceScheduleSelector();
+            ServiceSchedule__c actualSchedule = scheduleSelector.getServiceScheduleById(
+                expectedSchedule.Id
+            );
+            Test.stopTest();
+
+            System.assertEquals(
+                null,
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleSelector_TEST.cls:59]: The method signature is `public ServiceSchedule__c getServiceScheduleFromId(Id serviceScheduleId)` so a list would be an invalid return type

File: force-app/main/default/classes/ServiceScheduleService.cls (hunk 17 above)
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceScheduleService.cls:45]: I agree with the suggestion, but I'm debating whether it makes sense to tackle this change now or defer to the future.

File: force-app/main/default/classes/ServiceScheduleSelector_TEST.cls [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    private static ServiceScheduleSelector scheduleSelector = new ServiceScheduleSelector();
+
+    @TestSetup
+    private static void createScheduleData() {
+        TestDataFactory.generateServiceData();
+        TestDataFactory.generateServiceScheduleData('Daily', System.today());
+    }
+
+    @IsTest
+    private static void testgetServiceScheduleFromId() {
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleSelector_TEST.cls:24]: Typo: should be capital G

File: force-app/main/default/classes/ServiceScheduleSelector_TEST.cls [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+        System.assertEquals(
+            expectedSchedule.Name,
+            actualSchedule.Name,
+            'The Name of the returned record should match the record we requested'
+        );
+    }
+
+    @IsTest
+    private static void testGetServiceScheduleNoAccess() {
+        ServiceSchedule__c expectedSchedule = [SELECT Id FROM ServiceSchedule__c LIMIT 1];
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleSelector_TEST.cls:52]: Super nitpicky so feel free to ignore me but I would maybe name this existingSchedule since we no longer expect it to be returned.

File: force-app/main/default/lwc/serviceScheduleAddSessionsQuickAction/serviceScheduleAddSessionsQuickAction.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+ *  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ *
+ */
+
+import { LightningElement, api } from "lwc";
+import addServiceSessions from "@salesforce/label/c.Add_More_Sessions";
+
+export default class ServiceScheduleAddSessionsQuickAction extends LightningElement {
+    @api recordId;
+    @api isCommunity;
```
Comment (jjbennett) [inline force-app/main/default/lwc/serviceScheduleAddSessionsQuickAction/serviceScheduleAddSessionsQuickAction.js:15]: This is never set, thinking it is not needed. I do not believe we need to do anything special with quick actions in a community page. At least it didn't look like it on the other quick actions. Side note this is not showing up in the community at all right now.

File: force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js [hunk 23]
```diff
@@ -12,6 +12,7 @@ import newServiceSchedule from "@salesforce/label/c.New_Service_Schedule";
 
 export default class ServiceScheduleCreatorWrapper extends LightningElement {
     @api recordTypeId;
+    @api recordId;
```
Comment (jjbennett) [inline force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js:15]: Is this necessary?

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 24]
```diff
@@ -228,13 +312,31 @@ export default class NewServiceSchedule extends LightningElement {
...
         this.template.querySelectorAll("lightning-input-field").forEach(field => {
             if (field.fieldName === this.dateFields.start.apiName) {
-                this.setFirstSessionStartTimeAndEndTime(field.value);
+                let initialStartDate = new Date(field.value);
+                if (this.editingExistingSchedule) {
+                    let today = new Date();
+                    initialStartDate.setMonth(today.getMonth());
+                    initialStartDate.setDate(today.getDate());
+                    initialStartDate.setFullYear(today.getFullYear());
+                    console.log(initialStartDate);
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:327]: leftover console.log

File: force-app/main/default/staticresources/pmm/quickActionModal.css [hunk 25]
```diff
@@ -0,0 +1,4 @@
+.slds-modal__container {
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/staticresources/pmm/quickActionModal.css:1]: Scoping could be tricky - the component itself is nested a few layers WITHIN this container, so it may require some less-conventional css at least. Do you have that trick up your sleeve or should I go searching? 

Are these directives any stickier than the equivalent ones declared in the addMoreParticipants aura component?

[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/staticresources/pmm/quickActionModal.css:1]: Take a look at bsdt overrides

File: force-app/main/default/classes/ServiceSessionSelector.cls [hunk 26]
```diff
@@ -15,6 +15,26 @@ public with sharing class ServiceSessionSelector {
     @TestVisible
     private FieldBucketSelector bucketSelector = new FieldBucketSelector();
 
+    public List<ServiceSession__c> getFutureServiceSessionsForServiceSchedule(
+        Id serviceScheduleId
+    ) {
+        List<ServiceSession__c> serviceSessions = new List<ServiceSession__c>();
+
+        if (!Schema.SObjectType.ServiceSession__c.isAccessible()) {
```
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ServiceSessionSelector.cls:23]: I think I'm inclined to skip this one, both for pragmatically getting the WI closed and also because this matches the other methods in this class currently

File: force-app/main/default/aura/addMoreSessions/addMoreSessionsController.js [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+ *  * All rights reserved.
+ *  * SPDX-License-Identifier: BSD-3-Clause
+ *  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ *
+ */
+
+({
+    handleClose: function(component, event, helper) {
+        $A.get("e.force:closeQuickAction").fire();
+        $A.get("e.force:refreshView").fire();
```
Comment (jjbennett) [inline force-app/main/default/aura/addMoreSessions/addMoreSessionsController.js:12]: Another bonus with using Aura!

File: force-app/main/default/staticresources/pmm/quickActionModal.css (hunk 25 above)
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/staticresources/pmm/quickActionModal.css]: Switch to Aura no longer an issue.

File: force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js (hunk 23 above)
Comment (jjbennett) [reply force-app/main/default/lwc/serviceScheduleCreatorWrapper/serviceScheduleCreatorWrapper.js]: @ptewson-sfdo - this is the question I mentioned in slack.

**Review by jjbennett — APPROVED** [review]
